### PR TITLE
Add support for automatically starting a list of servers when the node starts

### PR DESCRIPTION
**Pull request by maffoo**
A longstanding point of pain when pushing out updates is the need to manually restart a bunch of servers after the node or the manager are restarted. This change adds an autostart list to the node which contains a list of servers that will be started automatically when the node connects (or reconnects) to the manager. Autostart can also be invoked manually, e.g. if you've stopped a bunch of servers and then want to start everything that is not yet running.

Note that I've lumped in a couple of other commits to this PR with minor changes like PEP8 stuff. Please take a look at the commits individually.

### CODE REVIEW

File: labrad/node/__init__.py [hunk 1]
```diff
@@ -548,14 +552,17 @@ def _load(self, nodename=None, create=False, defaults=None):
...
         p.get('directories', '*s', key='dirs')
         p.get('packages', '*s', key='mods')
         p.get('extensions', '*s', key='exts')
+        p.get('autostart', '*s', True, [], key='autostart')
         ans = yield p.send()
         dirs = filter(None, ans.dirs)
         mods = filter(None, ans.mods)
         exts = filter(None, ans.exts)
-        returnValue((dirs, mods, exts))
+        autostart = sorted(filter(None, ans.autostart))
```
Comment (DanielSank) [inline labrad/node/__init__.py]: That's a little obscure. I had to look up `filter` and then the special case where the first argument is `None`. List comprehension might be easier to understand.

File: labrad/node/__init__.py [hunk 2]
```diff
@@ -568,8 +575,19 @@ def _save(self):
...
-        self._update(config)
+        try:
+            config = yield self._load()
+            self._update(config)
+        except Exception:
+            logging.error('Error in _handleMessage', exc_info=True)
+
+    @inlineCallbacks
+    def update_autostart(self, autostart):
+        """Update the list of servers to be autostarted."""
```
Comment (DanielSank) [inline labrad/node/__init__.py]: Arg types, please.

File: labrad/node/__init__.py [hunk 3]
```diff
@@ -68,39 +68,42 @@
...
+
 LOG_LENGTH = 1000 # maximum number of lines of stdout to keep per server
 
+
 class IServerProcess(Interface):
     pass
 
+
 class ServerProcess(ProcessProtocol):
     """A class to represent a running server instance."""
```
Comment (DanielSank) [inline labrad/node/__init__.py]: I won't ask for a full doc string, but maybe just the attribute(s) added by this PR?

File: labrad/node/__init__.py [hunk 4]
```diff
@@ -835,6 +854,49 @@ def stream_output(self, c, name, enable):
...
+                yield deferred
+            except Exception:
+                logging.error('Failed to autostart "%s"', name, exc_info=True)
+
+    @setting(201, returns='*s')
+    def autostart_list(self, c):
+        """Get the list of servers that are configured to be autostarted."""
+        return self.config.autostart
+
+    @setting(202, name='s', returns='')
```
Comment (DanielSank) [inline labrad/node/__init__.py]: Consider returning bool indicating whether or not server was added, rather than raising exception.

**Comment by DanielSank** [comment]
Looks good (minor comments). I guess we ought to make sure it works first though.

**Comment by maffoo** [comment]
@DanielSank, please see latest changes. I've tested this locally, but let me know if you want to do any testing of your own before officially signing off.

**Comment by DanielSank** [comment]
Thanks for the docstrings. If you've tested that it works then I think we're good to go. I guess this kind of change isn't subject to the unit tests we normally require.

LGTM.

**Pull request merged**